src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
+                            worker.getColony().getID(),
+                            worker.getCitizenData().getId(),
+                            ((JobBuilder) job).getWorkOrderId()));
+                }
+                else
+                {
+                    if (!(wo instanceof WorkOrderBuildDecoration))
+                    {
+                        final AbstractBuilding building = job.getColony().getBuilding(wo.getBuildingLocation());
+                        if (building == null)
+                        {
+                            Log.getLogger().error(String.format("Builder (%d:%d) ERROR - Finished, but missing building(%s)",
+                                    worker.getColony().getID(),
+                                    worker.getCitizenData().getId(),
+                                    wo.getBuildingLocation()));
+                        }
+                        else
+                        {
+                            building.setBuildingLevel(wo.getUpgradeLevel());
+                        }
+                    }
+                    else
+                    {
+                        if (structureName.contains(WAYPOINT_STRING))
+                        {
+                            worker.getColony().addWayPoint(wo.getBuildingLocation(), world.getBlockState(wo.getBuildingLocation()));
+                        }
+                    }
+                    ((JobBuilder) job).complete();
+                }
+
+                final AbstractBuilding workerBuilding = getOwnBuilding();
+                if (workerBuilding instanceof BuildingBuilder)
+                {
+                    ((BuildingBuilder) workerBuilding).resetNeededResources();
+                }
+                resetTask();
+            }
+            else if(job instanceof JobMiner)
+            {
+                BuildingMiner minerBuilding = (BuildingMiner) getOwnBuilding();
+                //If shaft isn't cleared we're in shaft clearing mode.
+                if(!minerBuilding.clearedShaft)
+                {
+                    @NotNull final Level currentLevel = new Level(minerBuilding, ((JobMiner) job).getStructure().getPosition().getY());
+                    minerBuilding.addLevel(currentLevel);
+                    minerBuilding.setCurrentLevel(minerBuilding.getNumberOfLevels());
+                    minerBuilding.resetStartingLevelShaft();
+
+                    //Send out update to client
+                    getOwnBuilding().markDirty();
+                }
+                else
+                {
+                    minerBuilding.getCurrentLevel().closeNextNode(rotation);
+                }
+                ((JobMiner) job).setStructure(null);
+            }
+            worker.addExperience(XP_EACH_BUILDING);
+            workFrom = null;
+        }
+
+        currentStructure = null;
+
+        return AIState.IDLE;
+    }
+
+    private Boolean decorationStep(final Structure.StructureBlock structureBlock)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "decorationStep" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)